File: src/chal8.c
/* Detect AES in ECB mode
    In this file are a bunch of hex-encoded ciphertexts.

    One of them has been encrypted with ECB.

    Detect it.

    Remember that the problem with ECB is that it is stateless and deterministic; the same 16 byte plaintext block will always produce the same 16 byte ciphertext.

*/

#include <stdio.h>
#include <string.h>

#define CHUNK_SIZE 1024
char file_buf[CHUNK_SIZE];

// AES 128 bit block represented in hex takes 32 bytes 
#define BLOCK_LEN 32

int main() 
{
    // Loop through the whole file
    for (int line = 1; !feof(stdin); line++) {
        // Line by line
        fgets(file_buf, CHUNK_SIZE, stdin);
        // Get the length of the line read in
        size_t in_len = strnlen(file_buf, CHUNK_SIZE);
        // Loop through all the encrypted blocks
        for (int i = 0; i < in_len; i += BLOCK_LEN)
            // Loop through the remaining of the encrypted blocks
            for (int j = i + BLOCK_LEN; j < in_len; j += BLOCK_LEN)
                // Compare the two, print if a match is found
                if (memcmp(&file_buf[i], &file_buf[j], BLOCK_LEN) == 0)
                    printf("Row %d, Blocks %d, %d both %.32s\n", line, i/BLOCK_LEN, j/BLOCK_LEN, &file_buf[i]);
    }
    return 0;
}

File: chal8.py
""" Detect AES in ECB mode
    In this file are a bunch of hex-encoded ciphertexts.

    One of them has been encrypted with ECB.

    Detect it.

    Remember that the problem with ECB is that it is stateless and deterministic; the same 16 byte plaintext block will always produce the same 16 byte ciphertext.

"""
import base64
import itertools
import string

ciphertexts = []
with open("8.txt", 'r') as f:
    for line in f:
        ciphertexts.append(bytes.fromhex(line.strip()))

def grouper(n, iterable, fillvalue=None):
    """ grouper(3, 'ABCDEFG', 'x') --> ABC DEF Gxx """
    args = [iter(iterable)] * n
    return itertools.zip_longest(fillvalue=fillvalue, *args)

for i, ct in enumerate(ciphertexts):
    matches = 0
    for j, blk1 in enumerate(grouper(16, ct)):
        for k, blk2 in enumerate(grouper(16, ct)):
            if j != k and blk1 == blk2:
                print ("Row %d: Blocks %d, %d both %s" % (i + 1, j, k, bytes(blk1).hex()))
    